Python/Product/PythonTools/PythonTools/Intellisense/ExtractMethodSuggestedActionSourceProvider.cs

```diff
@@ -0,0 +1,206 @@
+using Microsoft.VisualStudio.Utilities;
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.PythonTools.Intellisense {
+    [Export(typeof(ISuggestedActionsSourceProvider))]
+    [Name("Python Extract Method Suggested Action")]
+    [ContentType(PythonCoreConstants.ContentType)]
+    [TextViewRole(PredefinedTextViewRoles.Analyzable)]
+    [TextViewRole(PredefinedTextViewRoles.Editable)]
+    class ExtractMethodSuggestedActionSourceProvider : ISuggestedActionsSourceProvider {
+        private readonly IServiceProvider _serviceProvider;
+        private readonly Lazy<PreviewChangesService> _changePreviewFactory;
+
+        [ImportingConstructor]
+        public ExtractMethodSuggestedActionSourceProvider([Import(typeof(SVsServiceProvider))] IServiceProvider provider, Lazy<PreviewChangesService> changePreviewFactory) {
+            _serviceProvider = provider;
+            _changePreviewFactory = changePreviewFactory;
+        }
+
+        public ISuggestedActionsSource CreateSuggestedActionsSource(ITextView textView, ITextBuffer textBuffer) {
+            return new ExtractMethodSuggestedActionSource(this, textView);
+        }
+
+        class ExtractMethodSuggestedActionSource : ISuggestedActionsSource {
+            private readonly ITextView _view;
+            private readonly ExtractMethodSuggestedActionSourceProvider _parent;
+            private bool? _canExtract;
+            private static Guid _telemetryId = new Guid("{0D887D91-1E11-4DC3-92F5-79A5182004C3}");
+
+            public ExtractMethodSuggestedActionSource(ExtractMethodSuggestedActionSourceProvider parent, ITextView view) {
+                _view = view;
+                _parent = parent;
+                _view.Selection.SelectionChanged += Selection_SelectionChanged;
+            }
+
+            private void Selection_SelectionChanged(object sender, EventArgs e) {
+                var newValue = MethodExtractor.CanExtract(_view);
+                if (newValue != _canExtract) {
+                    _canExtract = newValue;
+                    SuggestedActionsChanged?.Invoke(this, EventArgs.Empty);
+                }
+            }
+
+            public event EventHandler<EventArgs> SuggestedActionsChanged;
+
+            public void Dispose() {
+                _view.Selection.SelectionChanged -= Selection_SelectionChanged;
+            }
+
+            public IEnumerable<SuggestedActionSet> GetSuggestedActions(ISuggestedActionCategorySet requestedActionCategories, SnapshotSpan range, CancellationToken cancellationToken) {
+                if (MethodExtractor.CanExtract(_view) == true) {
+                    return new SuggestedActionSet[] {
+                    new SuggestedActionSet(
+                        new [] {  new SuggestedAction(_parent, _view) },
+                        SuggestedActionSetPriority.High
+                    )
+                };
+                }
+
+                return Array.Empty<SuggestedActionSet>();
```

Comment:
`Enumerable.Empty<T>()` (apparently faster due to some caching, but I'd be surprised if it wasn't negligible)